SPR1-3249-fix_check_applied_gain

tamerakassie opened the pull request:
In this commit we address the failing nmad phase plot generation due to sensor keys that are not available in the live telstate of an observation. We use aliases of `stream_name` `obs_label`  and `capture_block_id` to be used by `check_applied_gain` to make the required katdal object for us to get the applied gain values to check if the observation is phase up to trigger the pipeline to make nmad phase plots.

We also address the issue of the pipeline triggering nmad phase generation for image observations, this should not happen. Here we move the `bfcal` check as the first check, to flag and pass only observations that have the `bfcal` tag to the next applied gain check.

tamerakassie committed:
SPR1-3249-fix_check_applied_gain

In this commit we address the failing nmad phase plot generation due to
sensor keys that are not available in the live telstate of an
observation. We used found aliases of `stream_name', `obs_label` and
`capture_block_id` to be used by `check_applied_gain` to make the
required katdal object for us to get the applied gain values to check if
the observation is phase up to trigger the pipeline to make nmad phase plots.

We also address the issue of the pipeline triggering nmad phase
generation for image observations, this should not happen. Here we move
the `bfcal` check as the first check, to flag and pass only observations
that have the `bfcal` tag.

KimMcAlpine commented on katsdpcal/reduction.py:
```diff
@@ -135,7 +135,9 @@ def check_is_corrected(telstate, time_range):
         True for track that has a 'corrected' label in the `time_range`
     """
-    obs_label = telstate.get_range('obs_label', st=time_range[0], et=time_range[1],
+    capture_block_id = telstate.get_range('sdp_capture_block_id')[0][0]
```
It is safer/more predictable to give the `get_range` function the actual start and stop times of the track and you already have these as arguments to the function 
e.g. `telstate.get_range('sdp_capture_block_id', st=time_range[0], et=time_range[1])[0][0]`

KimMcAlpine commented on katsdpcal/reduction.py:
```diff
@@ -173,8 +175,8 @@ def check_applied_gain_sensor(telstate, ref_ant, pol):
 
-    capture_block_id = telstate['capture_block_id']
-    stream_name = telstate['stream_name']
+    capture_block_id = telstate.get_range('sdp_capture_block_id')[0][0]
```
Same as previous, give it the time range of the actual track to ensure you are selecting the correct cbid

KimMcAlpine commented on katsdpcal/reduction.py:
```diff
@@ -173,8 +175,8 @@ def check_applied_gain_sensor(telstate, ref_ant, pol):
+    stream_name = telstate['cal_src_streams'][0]
     telstate, capture_block_id, stream_name = view_l0_capture_stream(telstate.root(),
                                                                      capture_block_id, stream_name)
     source = TelstateDataSource(telstate, capture_block_id, stream_name, chunk_store=None)
```
As we discussed this results in a `chunk_info` not found error due to https://github.com/ska-sa/katdal/blob/0840fd86ca4954168cacf4cb785eb00afef121b4/katdal/datasources.py#L378 so you will need to include the timestamps as an argument here.

KimMcAlpine commented on katsdpcal/simulator.py:
```diff
@@ -125,7 +125,8 @@ def get_params(self):
             param_dict[key] = telstate[key]
 
-        telstate_mutables = ['cbf_target', 'obs_label', 'obs_activity']
+        telstate_mutables = ['cbf_target', 'obs_label', 'obs_activity', 'sdp_capture_block_id',
```
It seems like you have `obs_label` twice here once without the cbid prefix and once with ? Do you need both of these ? I think you can remove the `obs_label` one and just retain the `cbid_obs_label` sensor 


Also in your telstate_immutables list you now have some things that you don't need and won't have in a normal observation so I would remove these, these are `chunk_info`, `capture_block_id`, `stream_name`  and you will need to add `cal_src_streams` to the immutables list as you now access this sensor and it isn't included in the simulator.

KimMcAlpine replied:
you still need to remove 'chunk_info' here

KimMcAlpine commented on katsdpcal/simulator.py:
```diff
@@ -344,10 +345,11 @@ def setup_telstate(self, telstate):
         # separate keys without times from those with times
         sensor_key_suffixes = ('obs_activity', '_eq', 'cbf_target', 'target_activity', 'obs_label',
-                               'noise_diode')
+                               'noise_diode', 'sdp_capture_block_id',  f"{CBID}_obs_label")
```
You also have 'obs_label' twice here, you only need it once

KimMcAlpine reviewed (changes requested):
Check that all the sensors and attributes you need are in the simulated telstate and remove those that wouldn't ordinarily be there so that it doesn't confuse any future testing with the simulator.

KimMcAlpine commented on katsdpcal/reduction.py:
```diff
@@ -884,13 +886,13 @@ def pipeline(data, ts, parameters, solution_stores, stream_name, sensors=None):
-                if corrected_track:
-                    logger.info('Calculate NMAD on Corrected Track')
-                    if any(k in phase_tag for k in taglist):
+            if any(k in phase_tag for k in taglist):
```
It might be best to add a try, except option here so that if there are problems with this part of the code it doesn't prevent the rest of the pipeline from running and causing problems with the report making process.

tamerakassie committed:
SPR1-3249-fix_check_applied_gain

In this commit we added a try-except clause around the code that generates the phase nmad plots
We added the timestamps to build the katdal  data_source, this will prevent the failing due to unfound `chunk_info`
in some observations. We remove some unneed attributes/sensors from simulator.

KimMcAlpine commented on katsdpcal/simulator.py:
```diff
@@ -345,10 +345,10 @@ def setup_telstate(self, telstate):
             raise ValueError('number of substreams must divide into the number of channels')
         parameter_dict['sdp_l0_n_chans_per_substream'] = n_chans // self.n_substreams
-        CBID = parameter_dict['capture_block_id']
+        CBID = parameter_dict['sdp_capture_block_id'][1][0]
```
This won't work if there is only one element in the list, so consider `parameter_dict['sdp_capture_block_id'][-1][0]`

KimMcAlpine commented on katsdpcal/simulator.py:
```diff
@@ -116,17 +116,17 @@ def get_params(self):
+                               'obs_params', f'{correlator_stream}_{ins_name}',
+                               f'{correlator_stream}_int_time', f'{correlator_stream}_n_accs',
+                               f'{f_engine_stream}_{ins_name}', 'wide_scale_factor_timestamp',
+                               'wide_sync_time', 'chunk_info', 'cal_src_streams']
```
you still have 'chunk_info' here, remove this as it won't be present in a real telstate during an observation

KimMcAlpine commented on katsdpcal/reduction.py:
```diff
@@ -175,11 +177,15 @@ def check_applied_gain_sensor(telstate, ref_ant, pol):
     telstate, capture_block_id, stream_name = view_l0_capture_stream(telstate.root(),
                                                                      capture_block_id, stream_name)
-    source = TelstateDataSource(telstate, capture_block_id, stream_name, chunk_store=None)
+    timestamps = s.timestamps
```
I've realised using the scan timestamps won't actually work here, because now you are restricting sensor data for only the period of the scan, but the condition of changing the gain sensore twice applies to the whole phase-up script, ie within the scan it won't change twice. So you will need to construct all the timestamps for the observation here not just the scan.

tamerakassie replied:
I have added the following to build up the timestamps for the observation duration:
`t0 = telstate['sync_time'] + telstate['first_timestamp']
  int_time = telstate['int_time']
  chunk_info = telstate['chunk_info']
  n_dumps = chunk_info['correlator_data']['shape'][0]
  timestamps = t0 + np.arange(n_dumps) * int_time`

but I realize that here i do require `chunk_info` to build the number of dumps, the last time we checked/ran into the issue that this key `chuck_info` is not available in a live observation telstate. I might have to look at another way to make the timestamps, unless there is another way to get the number of dumps information.

tamerakassie committed:
Build timestamps for applied gain check

In this commit I have added a build for the entire duration timestamps based on the code
in `katdal` to pass the correct event occurances for the applied gain check.

Removed the last bit of `chunk_info` from the simulator.

Correctly index the sensor to extract the CBID.

tamerakassie committed:
Add chunk_info key that is available in a live ts

In this commit, I add the correct sensor key to get the n_dumps from
chunk_info, I tested in a live telstate that this is written as
CBID_sdp_l0_chunk_info.

tamerakassie committed:
added correct chunk info key name in live telstate

tamerakassie committed:
add correct timestamps build

tamerakassie commented:
@KimMcAlpine, I am putting a re-review for this work. I have check through my changes and I am happy with them.

tamerakassie committed:
Addressing PR comments

- Some tests of the phase nmad plots show that the phases for the h-pol
are cut off due to the ylim scaled by the standard deviation of the
data. We added the  mean + 5 * std as the scaling statistic for the ylim
of the phases instead.
- Cosmetic change : Set the last plot to only include the x-label (Frequency MHz) rather
  than both panels
